compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -341,7 +342,21 @@ impl<'tcx> InferCtxtExt<'tcx> for InferCtxt<'tcx> {
                     span: DUMMY_SP,
                     kind: TypeVariableOriginKind::MiscVariable,
                 });
-                let substs = self.tcx.mk_substs_trait(ty.skip_binder(), &[var.into()]);
+                let with_effects;
+                let without_effects;
+                let rest: &[_] = if self.tcx.effects() {
+                    let e_var = self.next_effect_var(
+                        DUMMY_SP,
+                        EffectVariableOriginKind::MiscVariable,
+                        ty::EffectKind::Constness,
+                    );
+                    with_effects = [var.into(), e_var.into()];
+                    &with_effects
+                } else {
+                    without_effects = [var.into()];
+                    &without_effects
+                };
+                let substs = self.tcx.mk_substs_trait(ty.skip_binder(), rest);
```

Comment:
`mk_substs_trait_with_effects`?

Reply:
We need to pick the host effect instead of something, so we can't use that. I did manage to clean it up a lot though.